File: lib/fx.ts
import "server-only";
import { Prisma, type PrismaClient } from "@prisma/client";
import { getLivePriceRows } from "@/lib/live-price";

// Quote-currency -> account-currency conversion (2026-09-23).
//
// A position's P&L comes out of computeRealizedPnl in the symbol's QUOTE currency ((close - open) x
// contractSize x volume: USDJPY in JPY, EURGBP in GBP, GER40 in EUR), and requiredMarginFor's
// volume x contractSize x price is a quote-currency notional too. Until this module both were added
// straight to the account's balance / margin as if they were already in the account's currency, which is
// only true when the two match (every XXXUSD symbol on a USD account). On USDJPY a 1-lot, 10-pip win booked
// 1,000 (JPY) as 1,000 USD, and its margin read ~150x too high. Every money figure now goes through
// conversionRate(quote, account) first; for the common case (same currency) that is exactly 1 and no price
// is read at all.
//
// Rate: the MID of the conversion pair's latest quote, as MT5 does for a cross when it has to pick one side
// for a figure that is not itself a fill. Tick-level freshness is NOT required: an FX rate a few minutes old
// moves a P&L by a fraction of a percent, while no conversion at all is wrong by the whole exchange rate.
// But an age LIMIT is (2026-09-25): a quote older than FX_RATE_MAX_AGE_MS counts as no price at all, so the
// caller refuses (FxRateUnavailableError / the position is unpriced) instead of converting with it. Before, a
// VPS outage fell back to Neon's LivePrice, which stopped moving when the feed's writes went VPS-local (S5):
// every non-USD conversion would silently have used a 10-day-old rate. 72 h still covers a normal weekend or
// holiday (Friday's FX close to the Sunday reopen is ~49 h: a BTC position in a EUR account is valued on
// Saturday with Friday's EURUSD, as MT5 does). The engine applies the same limit (order-management book.rs).

export type FxQuote = { bid: Prisma.Decimal; ask: Prisma.Decimal };
export type FxLookup = (symbol: string) => FxQuote | undefined;

const ONE = new Prisma.Decimal(1);

/** A conversion quote older than this is not used (see the header). Keep equal to the engine's FX_RATE_MAX_AGE. */
export const FX_RATE_MAX_AGE_MS = 72 * 3600 * 1000;

export class FxRateUnavailableError extends Error {
  constructor(public readonly from: string, public readonly to: string) {
    super(`no ${from}->${to} conversion rate: none of ${conversionSymbolsFor(from, to).join(", ")} has a price`);
  }
}

function mid(q: FxQuote | undefined): Prisma.Decimal | null {
  if (!q) return null;
  const m = q.bid.add(q.ask).div(2);
  return m.gt(0) ? m : null;
}

// One leg: from -> to through the pair itself or its inverse.
function direct(from: string, to: string, lookup: FxLookup): Prisma.Decimal | null {
  const d = mid(lookup(from + to));
  if (d) return d;
  const inv = mid(lookup(to + from));
  return inv ? ONE.div(inv) : null;
}

/** Multiply an amount in `from` by this to get `to`. null = no price to convert with. */
export function conversionRate(from: string, to: string, lookup: FxLookup): Prisma.Decimal | null {
  const f = from.trim().toUpperCase();
  const t = to.trim().toUpperCase();
  if (f === t) return ONE;
  const d = direct(f, t, lookup);
  if (d) return d;
  // cross through USD (JPY -> EUR = JPY -> USD -> EUR): every broker quotes the majors against USD
  if (f !== "USD" && t !== "USD") {
    const a = direct(f, "USD", lookup);
    const b = direct("USD", t, lookup);
    if (a && b) return a.mul(b);
  }
  return null;
}

/** Every symbol conversionRate may read for from -> to (none when they match). */
export function conversionSymbolsFor(from: string, to: string): string[] {
  const f = from.trim().toUpperCase();
  const t = to.trim().toUpperCase();
  if (f === t) return [];
  const out = [f + t, t + f];
  if (f !== "USD" && t !== "USD") out.push(f + "USD", "USD" + f, t + "USD", "USD" + t);
  return out;
}

type Db = PrismaClient | Prisma.TransactionClient;

/** Loads (in one read) the latest quotes every listed conversion may need. Pairs whose two currencies
 *  match need nothing, so the everyday all-USD case costs no query at all. */
export async function loadFxLookup(db: Db, pairs: Iterable<readonly [string, string]>): Promise<FxLookup> {
  const symbols = new Set<string>();
  for (const [from, to] of pairs) for (const s of conversionSymbolsFor(from, to)) symbols.add(s);
  if (symbols.size === 0) return () => undefined;
  const rows = await getLivePriceRows([...symbols], db);
  const cutoff = Date.now() - FX_RATE_MAX_AGE_MS;
  return (symbol) => {
    const r = rows.get(symbol);
    // too old to convert with = no price: refuse, never a stale rate
    return r && r.tickAt.getTime() > cutoff ? { bid: new Prisma.Decimal(r.bid), ask: new Prisma.Decimal(r.ask) } : undefined;
  };
}

/** The rate for one pair, loaded now. Throws FxRateUnavailableError when there is nothing to convert with. */
export async function quoteToAccountRate(db: Db, quoteCurrency: string, accountCurrency: string): Promise<Prisma.Decimal> {
  const lookup = await loadFxLookup(db, [[quoteCurrency, accountCurrency]]);
  const rate = conversionRate(quoteCurrency, accountCurrency, lookup);
  if (!rate) throw new FxRateUnavailableError(quoteCurrency.toUpperCase(), accountCurrency.toUpperCase());
  return rate;
}

File: engine/order-management/src/fx.rs
//! Quote-currency -> account-currency conversion (Rust cutover Stage 2, F2): a port of lib/fx.ts, same
//! rule, same test cases. A position's P&L and its margin notional come out in the symbol's QUOTE currency
//! (USDJPY in JPY); every money figure is multiplied by `conversion_rate(quote, account)` before it is added
//! to a balance, an equity or a used margin. Same currency is exactly 1 and needs no price.
//!
//! Rate = the MID of the conversion pair's latest quote: the pair itself, its inverse, or a cross through
//! USD. Freshness is not required (a few minutes' drift is a fraction of a percent; no conversion is wrong by
//! the whole exchange rate). A missing rate is None, and the caller treats that position as unpriced.

use rust_decimal::Decimal;

/// bid / ask of one symbol's latest quote
pub type Quote = (Decimal, Decimal);

fn mid(q: Option<Quote>) -> Option<Decimal> {
    let (bid, ask) = q?;
    let m = (bid + ask) / Decimal::TWO;
    (m > Decimal::ZERO).then_some(m)
}

fn direct<F: Fn(&str) -> Option<Quote>>(from: &str, to: &str, lookup: &F) -> Option<Decimal> {
    if let Some(d) = mid(lookup(&format!("{from}{to}"))) {
        return Some(d);
    }
    mid(lookup(&format!("{to}{from}"))).map(|inv| Decimal::ONE / inv)
}

/// Multiply an amount in `from` by this to get `to`. None = no price to convert with.
pub fn conversion_rate<F: Fn(&str) -> Option<Quote>>(from: &str, to: &str, lookup: F) -> Option<Decimal> {
    let f = from.trim().to_uppercase();
    let t = to.trim().to_uppercase();
    if f == t {
        return Some(Decimal::ONE);
    }
    if let Some(d) = direct(&f, &t, &lookup) {
        return Some(d);
    }
    if f != "USD" && t != "USD" {
        let a = direct(&f, "USD", &lookup)?;
        let b = direct("USD", &t, &lookup)?;
        return Some(a * b);
    }
    None
}

/// Every symbol `conversion_rate` may read for from -> to (none when they match).
pub fn conversion_symbols_for(from: &str, to: &str) -> Vec<String> {
    let f = from.trim().to_uppercase();
    let t = to.trim().to_uppercase();
    if f == t {
        return Vec::new();
    }
    let mut out = vec![format!("{f}{t}"), format!("{t}{f}")];
    if f != "USD" && t != "USD" {
        out.extend([format!("{f}USD"), format!("USD{f}"), format!("{t}USD"), format!("USD{t}")]);
    }
    out
}

/// A realized P&L in the account currency, rounded exactly like lib/position-close.ts: unchanged when the rate
/// is 1, otherwise to 4 dp half away from zero (decimal.js ROUND_HALF_UP, what Prisma.Decimal uses).
pub fn convert_pnl(quote_pnl: Decimal, rate: Decimal) -> Decimal {
    if rate == Decimal::ONE {
        quote_pnl
    } else {
        (quote_pnl * rate).round_dp_with_strategy(4, rust_decimal::RoundingStrategy::MidpointAwayFromZero)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rust_decimal_macros::dec;
    use std::collections::HashMap;

    // lib/fx.test.ts, case for case
    fn book(quotes: &[(&str, Decimal, Decimal)]) -> impl Fn(&str) -> Option<Quote> {
        let m: HashMap<String, Quote> = quotes.iter().map(|(s, b, a)| (s.to_string(), (*b, *a))).collect();
        move |s| m.get(s).copied()
    }

    #[test]
    fn same_currency_is_exactly_one() {
        assert_eq!(conversion_rate("USD", "usd", book(&[])), Some(Decimal::ONE));
    }

    #[test]
    fn direct_pair_mid() {
        assert_eq!(conversion_rate("GBP", "USD", book(&[("GBPUSD", dec!(1.35000), dec!(1.35020))])), Some(dec!(1.3501)));
    }

    #[test]
    fn inverse_pair() {
        let r = conversion_rate("JPY", "USD", book(&[("USDJPY", dec!(149.990), dec!(150.010))])).unwrap();
        assert_eq!((r * dec!(150)).round_dp(10), Decimal::ONE);
    }

    #[test]
    fn cross_through_usd() {
        let r = conversion_rate("JPY", "EUR", book(&[("USDJPY", dec!(150), dec!(150)), ("EURUSD", dec!(1.2), dec!(1.2))])).unwrap();
        assert_eq!(r.round_dp(12), (Decimal::ONE / dec!(150) / dec!(1.2)).round_dp(12));
    }

    #[test]
    fn nothing_to_convert_with_and_zero_quote() {
        assert_eq!(conversion_rate("JPY", "USD", book(&[])), None);
        assert_eq!(conversion_rate("JPY", "USD", book(&[("USDJPY", dec!(0), dec!(0))])), None);
    }

    #[test]
    fn usdjpy_ten_pip_win_is_usd() {
        let pnl_jpy = (dec!(150.100) - dec!(150.000)) * dec!(100000);
        let r = conversion_rate("JPY", "USD", book(&[("USDJPY", dec!(150.000), dec!(150.000))])).unwrap();
        assert_eq!((pnl_jpy * r).round_dp(2), dec!(66.67));
        assert_eq!(convert_pnl(pnl_jpy, r), dec!(66.6667));
    }

    #[test]
    fn symbols_it_may_read() {
        assert!(conversion_symbols_for("USD", "USD").is_empty());
        assert_eq!(conversion_symbols_for("JPY", "USD"), vec!["JPYUSD", "USDJPY"]);
        assert_eq!(conversion_symbols_for("JPY", "EUR"), vec!["JPYEUR", "EURJPY", "JPYUSD", "USDJPY", "EURUSD", "USDEUR"]);
    }

    #[test]
    fn convert_pnl_rounds_half_away_from_zero_like_decimal_js() {
        assert_eq!(convert_pnl(dec!(-150000), Decimal::ONE / dec!(148.5)), dec!(-1010.1010));
        assert_eq!(convert_pnl(dec!(0.00005), dec!(2)), dec!(0.0001)); // 0.0001 exactly, no rounding
        assert_eq!(convert_pnl(dec!(0.000025), dec!(2)), dec!(0.0001)); // 0.00005 -> away from zero
        assert_eq!(convert_pnl(dec!(123.45), Decimal::ONE), dec!(123.45)); // rate 1: untouched
    }
}
